File: 1235.maximum-profit-in-job-scheduling/solution.py
class Solution:
    def jobScheduling(self, startTime: List[int], endTime: List[int], profit: List[int]) -> int:
        N = len(startTime)
        
        jobs = [
            {
                "start": startTime[i],
                "end": endTime[i],
                "profit": profit[i],
            } for i in range(N)
        ]
        
        # ジョブを開始時間順に並び替える
        jobs.sort(key = cmp_to_key(lambda x, y: x["start"] - y["start"]))
                  
        dp = {}
        def findNextJobIndex(currentEndTime: int) -> int:
            start = 0
            end = N - 1
            idx = N
            while (start <= end):
                mid = (start + end) >> 1
                if (jobs[mid]["start"] >= currentEndTime):
                    end = mid - 1
                    idx = mid
                else:
                    start = mid + 1
            return idx
            
        def backtrack(position: int) -> int:
            if (position == N):
                return 0
                  
            if (position in dp):
                return dp[position]
            
            # jobs[position] を選択しない場合
            doNothing = backtrack(position + 1)
            
            # jobs[position] を選択するとして、次の開始位置を探す
            nextIdx = findNextJobIndex(jobs[position]["end"])
            
            # jobs[nextIdx] から探索を始める
            doJob = backtrack(nextIdx) + jobs[position]["profit"]
            
            dp[position] = max(doNothing, doJob)
            return dp[position]
        
        return backtrack(0)

File: 1235.maximum-profit-in-job-scheduling/solution.ts
type Job = {
  start: number;
  end: number;
  profit: number;
};

function findNextJobIndex(jobs: Job[], start: number, prevJobEndTime: number): number {
  let end = jobs.length - 1;

  while (start <= end) {
      const mid = (start + end) >>> 1;
      if (jobs[mid].start < prevJobEndTime) {
          start = mid + 1;
      } else {
          end = mid - 1;
      }
  }
  return start;
}

function jobScheduling(startTime: number[], endTime: number[], profit: number[]): number {
  const numOfJobs = startTime.length;
  
  const jobs: Job[] = [];
  for (let i = 0; i < numOfJobs; i++) {
      jobs.push({
          start: startTime[i],
          end: endTime[i],
          profit: profit[i],
      });
  }
  jobs.sort((a, b) => a.start - b.start || a.end - b.end);
  

  const dp = new Array(numOfJobs + 1).fill(0);

  for (let i = 0; i < numOfJobs; i++) {
      const nextJobIdx = findNextJobIndex(jobs, i, jobs[i].end);
      dp[nextJobIdx] = Math.max(dp[nextJobIdx], dp[i] + jobs[i].profit);
      dp[i + 1] = Math.max(dp[i + 1], dp[i])
  }
  return dp[numOfJobs];
};